articles/events_executor.md

```diff
@@ -129,11 +132,63 @@ This class will forward the items needed for pushing events to the notify guard
+
+Consider the following example of a system with two subscriptions A and B. Subscription A has a QoS history size of 1.
+While the `EventsExecutor` is busy processing events, the following events accumulates into the `EventsQueue`:
+ - Event 1 from Subscription A
+ - Event 2 from Subscription B
+ - Event 3 from Subscription A
+
+When the `EventsExecutor` is ready to process these new events, it will start from the first pushed, i.e. the event 1 from Subscription A.
+However, when calling the `execute_subscription()` API, the message that generated event 1 will not be available anymore as it has been overwritten from the message that generated event 3.
+The `execute_subscription()` API will not fail, but rather it will process the message that generated event 3.
+This violates the ordering as event 2 should have been processed before that.
+
+##### LockFreeEventsQueue
+
+This queue is equivalent to the `SimpleEventsQueue` with the difference that is lock free, which improves performance reducing times needed to enqueue/dequeue events from it.
+
+##### BoundedEventsQueue
+
+This queue doesn't allow more events from an entity than its history size. For example a subscription with a history size of 5, can't have more than 5 events from it in the queue.
+
+This queue has policies to decide what to do when a new event arrives from an entity which will exceed the amount of events allowed. It can remove the oldest event and push a new one, which keeps the relative time ordering of the event with respect to other events (at the cost of some CPU time), or it can directly avoid pushing the new event into the queue, saving CPU but subverting the time ordering of events.
+
+##### WaitSetEventsQueue
+
+This queue has a waitset-like behaviour: There's a single entry for each entity, and the order of events execution is the same as the `SingleThreadedExecutor` executor.
+If the `TimersManager` is configured to push timer events, timers will be executed in the same thread as other entities, as it happens also with the current default executor.
+
+The difference with respect to the waitset used in the `SingleThreadedExecutor` is that only entities which have work to do are present in the waitset. This way we avoid polling entities in search of new data to process.
+Also this reduces the time to look for existing elements in the waitset (to update their events counter) as there are less elements to iterate.
+
+The events from a single entity are represented by a single event and counter, specifying the number of events left to execute from this entity (as oposed to other EventsQueue which can have multiple events).
+
+The waitset has elements ordered by entity tipe: 1.Timers / 2.Subs / 3.Services / 4.Clients / 5.Waitables. An iterator points to the next event to dequeue, so the order of execution of entities is the same as of the `SingleThreadedExecutor`.
```

Comment:
all these are implementation details, they are not relevant

Reply:
Removed in 0d70bf2